api-authentication/server/index.js

```diff
@@ -0,0 +1,108 @@
+require('dotenv/config');
+const pg = require('pg');
+const argon2 = require('argon2'); // eslint-disable-line
+const express = require('express');
+const jwt = require('jsonwebtoken'); // eslint-disable-line
+const ClientError = require('./client-error');
+const errorMiddleware = require('./error-middleware');
+
+const db = new pg.Pool({ // eslint-disable-line
+  connectionString: process.env.DATABASE_URL,
+  ssl: {
+    rejectUnauthorized: false
+  }
+});
+
+const app = express();
+
+const jsonMiddleware = express.json();
+
+app.use(jsonMiddleware);
+
+app.post('/api/auth/sign-up', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(400, 'username and password are required fields');
+  }
+  argon2
+    .hash(password)
+    .then(hashedPassword => {
+      const sql = `
+        insert into "users" ("username", "hashedPassword")
+        values ($1, $2)
+        returning "userId", "username", "createdAt"
+      `;
+      const params = [username, hashedPassword];
+      return db.query(sql, params);
+    })
+    .then(result => {
+      const [user] = result.rows;
+      res.status(201).json(user);
+    })
+    .catch(err => next(err));
+});
+
+app.post('/api/auth/sign-in', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(401, 'invalid login');
+  }
+
+  /* your code starts here */
+  const sql = ` select "userId",
+  "hashedPassword",
+  "username"
+  from "users"
+  where "username" = $1
+  `;
+  const value = [username];
+  db.query(sql, value)
+    .then(result => {
+      const rows = result.rows[0];
+      if (result.rows.length === 0) {
+        throw new ClientError(401, 'invalid login');
+      }
+      argon2.verify(result.rows[0].hashedPassword, password)
+        .then(isPassword => {
+          if (!isPassword) {
+            throw new ClientError(401, 'invalid login');
+          }
+          const payload = {
+            userId: rows.userId,
+            username: rows.username
+          };
+          const token = jwt.sign(payload, process.env.TOKEN_SECRET);
+          res.status(201).json({
```

Comment:
A status `200` would be more appropriate here since nothing is being created.

Reply:
Will change for future reference thank you!